lib/jsonrpc/client.go

```diff
@@ -173,13 +175,28 @@ func (c *client) makeOutChan(ctx context.Context, ftyp reflect.Type, valOut int)
 			go func() {
 				for buf.Len() > 0 {
 					front := buf.Front()
-
 					bufLk.Unlock()
 
-					ch.Send(front.Value.(reflect.Value).Elem()) // todo: select on ctx is probably a good idea
-
-					bufLk.Lock()
-					buf.Remove(front)
+					cases :=  []reflect.SelectCase{
+						{
+							Dir: reflect.SelectRecv,
+							Chan: reflect.ValueOf(chCtx.Done()),
+						},
+						{
+							Dir:  reflect.SelectSend,
+							Chan: ch,
+							Send: front.Value.(reflect.Value).Elem(),
+						},
+					}
+
+					switch chosen, _, _ := reflect.Select(cases); chosen {
+					case 0:
+						bufLk.Lock()
```

Comment:
I'd move this before the switch (with the select), this way we only call Lock once

Reply:
but it should lock after channel send.

Reply:
Just put `chosen, _, _ := reflect.Select(cases)` before the loop as well

Reply:
then how to get front value?

Reply:
oh.I Got it.